Build a regression model.

In [3]:
#importing packages 
import pandas as pd
import statsmodels.api as sm

file_path_poi = '/Users/priyaganesan/Desktop/python_exercise/Pandas_exercise/db/Final-Project-statistical-Modelling-with-Python/poi_csv.csv'
file_path_vegrest='/Users/priyaganesan/Desktop/python_exercise/Pandas_exercise/db/Final-Project-statistical-Modelling-with-Python/vegrest_csv.csv'
file_path_library='/Users/priyaganesan/Desktop/python_exercise/Pandas_exercise/db/Final-Project-statistical-Modelling-with-Python/library_csv.csv'
#Read the CSV file into a DataFrame

poi_df = pd.read_csv(file_path_poi)
vegrest_df=pd.read_csv(file_path_vegrest)
library_poi=pd.read_csv(file_path_library)

Provide model output and an interpretation of the results. 

In [4]:
# loading dataset
vegrest_df

name  review  rating            categories         phone  \
0            Udupi Palace     208     3.5  vegetarianrestaurant  1.416406e+10   
1            KaSpace Cafe       9     5.0  vegetarianrestaurant  1.647658e+10   
2        Bombay Chowpatty      59     3.0  vegetarianrestaurant  1.416406e+10   
3     The Social Gardener       5     3.5  vegetarianrestaurant  1.416466e+10   
4            Udupi Palace     208     3.5  vegetarianrestaurant  1.416406e+10   
...                   ...     ...     ...                   ...           ...   
2479         Fit For Life      10     2.5  vegetarianrestaurant  1.416363e+10   
2480              Freshii      11     3.0  vegetarianrestaurant  1.647693e+10   
2481         Fit For Life       1     4.0  vegetarianrestaurant  1.416369e+10   
2482              Freshii       2     3.5  vegetarianrestaurant  1.647347e+10   
2483              Freshii       6     2.0  vegetarianrestaurant  1.647485e+10   

         distance                                            address  \
0      782.410965  1432 Gerrard Street East, Toronto, ON M4L 1Z6,...   
1     1023.018862  1432 Gerrard Street East, Toronto, ON M4L 1Z6,...   
2      832.357744  1432 Gerrard Street East, Toronto, ON M4L 1Z6,...   
3      839.372147  1432 Gerrard Street East, Toronto, ON M4L 1Z6,...   
4      750.617283  1432 Gerrard Street East, Toronto, ON M4L 1Z6,...   
...           ...                                                ...   
2479   656.863144  100 Queen Street W, Nathan Phillips Square, To...   
2480  1021.387572  100 Queen Street W, Nathan Phillips Square, To...   
2481   762.282582  100 Queen Street W, Nathan Phillips Square, To...   
2482   561.771465  100 Queen Street W, Nathan Phillips Square, To...   
2483   790.938858  100 Queen Street W, Nathan Phillips Square, To...   

                           station_name  free_bikes  
0             Queen St E / Woodward Ave         7.0  
1             Queen St E / Woodward Ave         7.0  
2             Queen St E / Woodward Ave         7.0  
3             Queen St E / Woodward Ave         7.0  
4               Queen St E / Rhodes Ave        20.0  
...                                 ...         ...  
2479  Richmond St E / Jarvis St Green P        11.0  
2480  Richmond St E / Jarvis St Green P        11.0  
2481  Richmond St E / Jarvis St Green P        11.0  
2482  Richmond St E / Jarvis St Green P        11.0  
2483  Richmond St E / Jarvis St Green P        11.0  

[2484 rows x 9 columns]

Build a regression model using Python’s statsmodels module that demonstrates a relationship between the number of bikes in a particular location and the characteristics of the POIs in that location.(rating,distance)
null hypothesis :  'rating' and 'distance'  has no effect on the number of free bikes. 



In [16]:
# Defining the independent variable and the dependent variable

X = poi_df[['distance', 'rating']] # independent variables
y = poi_df[['free_bikes']] # dependent variable

# Add a contant term to the independent variables (intercept)
X = sm.add_constant(X)

# create and fit the linear regression model
model = sm.OLS(y, X).fit()

In [17]:
# Print a summary of the regression results
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.024
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     45.04
Date:                Tue, 12 Dec 2023   Prob (F-statistic):           4.74e-20
Time:                        20:39:53   Log-Likelihood:                -13042.
No. Observations:                3677   AIC:                         2.609e+04
Df Residuals:                    3674   BIC:                         2.611e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.9342      0.708     21.079      0.0

Model Summary:

 R-squared is 0.024 and Adj. R-squared is 0.023.Our multivariate model explains 2.3% of the variations in the data.
All p values are 0 and < 0.05. This indicates that we can reject null hypothesis and there is correlation between free bikes and distance and rating
distance has a coef of -0.0036  and rating has a coef of -0.8789 
distance:So, if 'distance' is measured in meters, the negative coefficient suggests that as the distance from a location increases by one meter, the model predicts a small decrease in the number of free bikes. 
rating:rating has a coef of -0.8789 
Therefore, the 'rating' variable is considered statistically significant in predicting the number of free bikes in this model.
In this case, the model suggests that, on average, locations with higher ratings tend to have fewer free bikes, but it doesn't provide information about the underlying reasons for this relationship. Additional analysis and domain knowledge may be necessary to understand the implications of the observed relationships in this specific dataset.

# Stretch

How can you turn the regression model into a classification model?

1)Decide on a threshold value Create Categorical Labels:
2)Introduce a new categorical variable, say 'Availability Category,' where stations with free bikes above the threshold are labeled as "High" and others as "Low."
3)Update your dataset to include these new categories
4)Select a classification algorithm suitable for your dataset and problem. Common choices include logistic regression, decision trees, or random forests.
